include/spirv-tools/libspirv.hpp

```diff
@@ -31,6 +31,10 @@ using MessageConsumer = std::function<void(
     const spv_position_t& /* position */, const char* /* message */
     )>;
 
+// Create a context with the targeted environemnt |env| and set its message
```

Comment:
This seems naked.  It provides the creation idiom but not the destruction idiom.  The complexity is exposed in that you have to manually call spvContextDestroy in all the exit paths.

It would be better to have a RAII class that is convertible to an spv_context reference.  E.g.:

```
class Context {
  public:
   Context(spv_target_env  e, MessageConsumer mc);
   Context(Context&& other);
   ~Context() { spvContextDestroy(c); }
   spv_context&  operator() { return context_; }
  private:
   spv_context context_;
}
```


Inside a .cpp file:
```
   Context::Context(spv_target_env e, MessageConsumer mc) { 
     context_ = spvContextCreate(e); SetContextMessageConsumer(context_, mc); 
   }
   Context::Context(Context&& other)  { context_ = other.context_; other.context_ = nullptr; }
```

Then in linker.cpp you can create a Context object and pass it in as the argument where spv_context& is expected.

Reply:
Also the Context's copy constructor and assignment operators should be deleted.

Reply:
That would be better, indeed. Another option would be to pass an `spv_diagnostic*` to `Link()` to make it more similar to other entry points like `spvValidate()`, `spvBinaryParse()`, etc. It would be more C-friendly, and I could even get rid of `std::vector<>` and have it in “libspirv.h”.

Reply:
Which option would you prefer (between going with the C-API, with spv_diagnostic, or the Context class)? @dneto0

Reply:
I prefer the Context class for now.  The spv_diagnostic API design does not really work when you might emit more than one message.  The message consumer is better.
(But I also regret that that is such a complex lambda!  It should just have taken a string or string-view)

If someone wants a C API then we can make a C API version that wraps around the C++ version.  :-)

Thanks for asking.

Reply:
Sounds good, I’ll add the Context class. :-)